File: automation/klingenregel.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
klingenregel.py — EINE Quelle für die Klingen-Hausregel (Google-Kanal).

Vorher stand dieselbe Regex wörtlich in fünf Dateien (drei CJ-Importer, zwei
Google-Kanal-Wächter). Am 28.08. musste sie in allen fünf repariert werden, am
29.08. erneut — das ist die sechste Wiederholung der Geschwister-Lehre nach
Farbtabelle, Grössenmenge, publishVerified(), Preisformel und technik_plausibel.
**Neue Klingenwörter NUR in `klingenregel.json`.**

Die Regel hat drei Stufen, und jede stammt aus einem echten Fehlgriff:

1. `klinge_am_ende` — trifft die Ware SELBST. Die Wortgrenze steht am ENDE, weil
   deutsche Zusammensetzungen das Grundwort hinten tragen («Küchenmesser»).
   Zubehör wie «Messerblock» fällt damit korrekt heraus: Küchenbesteck und
   Zubehör sind bei Google zulässig (Hausregel 12.08.).

2. `waffenwort_vorne` — schliesst die Lücke vom 29.08.: «Retro Schwertabdeckung»
   ist eine Scheide für ein Schwert und fiel durch Stufe 1, weil das Waffenwort
   VORNE steht. Enthält bewusst KEIN «messer» — sonst fielen Messerblock und
   Messerschärfer wieder heraus.

3. `waffenwort_vorne_ausnahme` — die Gegenrichtung: «Schwertblatt» ist der
   Bogenhanf, «Schwertwal» der Orca, «Schwertlilie» die Iris. Ohne diese Liste
   hätte Stufe 2 ein Deko-Kissen aus dem einzigen verkaufenden Kanal geworfen.

Nutzung:  from klingenregel import ist_klinge
          if ist_klinge(titel): ...
"""
import json
import os
import re

_PFAD = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'klingenregel.json')
_R = json.load(open(_PFAD, encoding='utf-8'))

KLINGE_AM_ENDE   = re.compile(_R['klinge_am_ende'], re.I)
WAFFENWORT_VORNE = re.compile(_R['waffenwort_vorne'], re.I)
VORNE_AUSNAHME   = re.compile(_R['waffenwort_vorne_ausnahme'], re.I)
MESSGERAET       = re.compile(_R['messgeraet_ausnahme'], re.I)
KONTEXT_AUSNAHME = re.compile(_R['kontext_ausnahme'], re.I)
HANDKLINGE_STAMM = re.compile(_R['handklinge_stamm'], re.I)
HANDKLINGE_KEIN_PAKET = re.compile(_R['handklinge_kein_paket'], re.I)
HANDKLINGE_GERAET = re.compile(_R['handklinge_geraet'], re.I)
HANDKLINGE_SPIELZEUG = re.compile(_R['handklinge_spielzeug'], re.I)
HANDKLINGE_MIT_ZUBEHOER = re.compile(_R['handklinge_mit_zubehoer'], re.I)
_KLINGEN_NOMEN = ('messer', 'messern', 'knife', 'knives')


def ist_klinge(titel):
    """True, wenn der Titel eine Klinge oder Waffenzubehör bezeichnet.

    Die Kontext-Ausnahme steht ZUERST: «Washed Machete Jeans» ist eine Waschung,
    «Samurai mit Katana» eine Dekofigur. Ein Kleidungs- oder Deko-Wort im Titel
    sticht die Klingenwörter — sonst räumt die Regel halbe Abteilungen leer.
    """
    t = titel or ''
    if KONTEXT_AUSNAHME.search(t):
        return False
    if KLINGE_AM_ENDE.search(t) and not MESSGERAET.search(t):
        return True
    if WAFFENWORT_VORNE.search(t) and not VORNE_AUSNAHME.search(t):
        return True
    return False


def ist_handklinge(titel):
    """True, wenn im PAKET eine Handklinge liegt — die Versandfrage, nicht die Werbefrage.

    Getrennt von `ist_klinge`, weil es eine andere Frage ist (Lehre 16.09.2026,
    Bestellung #1017): CJ hat die Sendung aus Shanghai als VERBOTENEN ARTIKEL
    zurueckgeschickt. Fuer die Schweiz gibt es keine Linie fuer Klingen — egal ob
    Kuechenmesser oder Taschenmesser, und egal was `freightCalculate` vorher sagte.

    Zwei Richtungen, in denen sich die Antwort von `ist_klinge` unterscheidet:

    WEITER: «Edelstahl-Messerset» und «Keramikmesser-Set» sind fuer `ist_klinge`
    FALSE — deren Endverankerung scheitert am Fugen-s bzw. am Plural-n. Fuer die
    Versandfrage zaehlt der Wortstamm, denn im Karton liegen Klingen.

    ENGER: Zubehoer ohne Klinge (Messerblock, -halter, -schaerfer, Schwertpflegeoel,
    Schwertgurt) darf weiter in die Schweiz und bleibt im Verkauf. Ebenso Geraete
    mit gekapselter Klinge (Standmixer, Rasierer, Schaeler, Schere) und die
    Messgeraete-Ausnahme («Handkraft-Messer» ist ein Dynamometer).

    Die Kontext-Ausnahme gilt wie bei `ist_klinge` zuerst: «Washed Machete Jeans».
    """
    t = titel or ''
    if KONTEXT_AUSNAHME.search(t):
        return False
    if HANDKLINGE_SPIELZEUG.search(t):
        return False
    if MESSGERAET.search(t):
        return False
    if WAFFENWORT_VORNE.search(t) and VORNE_AUSNAHME.search(t):
        return False
    # 23.09.2026: Paket MIT Klinge — «Hackmesser mit Schutzhülle», «Messerblock mit 6 Messern».
    # Steht VOR der Zubehör-Ausnahme, weil «schutzhülle»/«scheide» dort als reines Zubehör gelten.
    m = HANDKLINGE_MIT_ZUBEHOER.search(t)
    if m and not HANDKLINGE_GERAET.search(t):
        vor = t[:m.start()]
        if m.group(1).lower() in _KLINGEN_NOMEN:
            return True
        if HANDKLINGE_STAMM.search(vor) and not HANDKLINGE_KEIN_PAKET.search(vor):
            return True
    if HANDKLINGE_KEIN_PAKET.search(t) or HANDKLINGE_GERAET.search(t):
        return False
    return bool(HANDKLINGE_STAMM.search(t))

File: automation/klingenregel.mjs
// klingenregel.mjs — EINE Quelle für die Klingen-Hausregel (Google-Kanal).
//
// Vorher stand dieselbe Regex wörtlich in fünf Dateien (cj_category_fill.mjs,
// cj_sku_import.mjs, cj_trending_import.mjs, google_kanal_luecke.py,
// google_kanal_luecke_schliessen.py). Am 28.08. musste sie in allen fünf
// repariert werden, am 29.08. erneut. Sechste Wiederholung der Geschwister-
// Lehre nach Farbtabelle, Grössenmenge, publishVerified(), Preisformel und
// technik_plausibel. **Neue Klingenwörter NUR in klingenregel.json.**
//
// Die JSON liegt neben dieser Datei; sie wird über import.meta.url aufgelöst,
// damit der Aufrufer aus jedem Arbeitsverzeichnis starten kann (die Runner
// laufen aus /tmp).
//
// Drei Stufen, jede aus einem echten Fehlgriff:
//  1. klinge_am_ende      — die Ware selbst. Wortgrenze am ENDE, weil deutsche
//                           Zusammensetzungen das Grundwort hinten tragen.
//                           «Messerblock»/«Messerschärfer» bleiben damit
//                           korrekt draussen (Zubehör ist bei Google zulässig).
//  2. waffenwort_vorne    — Lücke 29.08.: «Retro Schwertabdeckung» ist eine
//                           Scheide und fiel durch Stufe 1. Ohne «messer», sonst
//                           fiele der Messerblock wieder heraus.
//  3. *_ausnahme          — Gegenrichtung: «Schwertblatt» ist der Bogenhanf,
//                           «Schwertwal» der Orca.
import { readFileSync } from 'node:fs';

const R = JSON.parse(readFileSync(new URL('./klingenregel.json', import.meta.url), 'utf8'));

export const KLINGE_AM_ENDE   = new RegExp(R.klinge_am_ende, 'i');
export const WAFFENWORT_VORNE = new RegExp(R.waffenwort_vorne, 'i');
export const VORNE_AUSNAHME   = new RegExp(R.waffenwort_vorne_ausnahme, 'i');
export const MESSGERAET       = new RegExp(R.messgeraet_ausnahme, 'i');
export const KONTEXT_AUSNAHME = new RegExp(R.kontext_ausnahme, 'i');
export const HANDKLINGE_STAMM      = new RegExp(R.handklinge_stamm, 'i');
export const HANDKLINGE_KEIN_PAKET = new RegExp(R.handklinge_kein_paket, 'i');
export const HANDKLINGE_GERAET     = new RegExp(R.handklinge_geraet, 'i');
export const HANDKLINGE_SPIELZEUG  = new RegExp(R.handklinge_spielzeug, 'i');
export const HANDKLINGE_MIT_ZUBEHOER = new RegExp(R.handklinge_mit_zubehoer, 'i');
const KLINGEN_NOMEN = new Set(['messer', 'messern', 'knife', 'knives']);

// Die Kontext-Ausnahme steht ZUERST: «Washed Machete Jeans» ist eine Waschung,
// «Samurai mit Katana, 30 cm» eine Dekofigur. Ein Kleidungs- oder Deko-Wort
// sticht die Klingenwörter — sonst räumt die Regel halbe Abteilungen leer.
export function istKlinge(titel) {
  const t = titel || '';
  if (KONTEXT_AUSNAHME.test(t)) return false;
  if (KLINGE_AM_ENDE.test(t) && !MESSGERAET.test(t)) return true;
  if (WAFFENWORT_VORNE.test(t) && !VORNE_AUSNAHME.test(t)) return true;
  return false;
}

// istHandklinge — die VERSANDFRAGE, nicht die Werbefrage (Lehre 16.09.2026, #1017).
// CJ hat die Sendung aus Shanghai als VERBOTENEN ARTIKEL zurueckgeschickt; fuer die
// Schweiz gibt es keine Linie fuer Klingen, egal was freightCalculate vorher sagte.
// WEITER als istKlinge: «Edelstahl-Messerset» faellt dort durchs Fugen-s, hier nicht.
// ENGER als istKlinge: Zubehoer ohne Klinge im Paket (Block, Halter, Schaerfer,
// Schwertpflegeoel) und Geraete mit gekapselter Klinge (Mixer, Rasierer, Schere)
// duerfen weiter in die Schweiz und bleiben im Verkauf.
export function istHandklinge(titel) {
  const t = titel || '';
  if (KONTEXT_AUSNAHME.test(t)) return false;
  if (HANDKLINGE_SPIELZEUG.test(t)) return false;
  if (MESSGERAET.test(t)) return false;
  if (WAFFENWORT_VORNE.test(t) && VORNE_AUSNAHME.test(t)) return false;
  // 23.09.2026: Paket MIT Klinge — «Hackmesser mit Schutzhülle», «Messerblock mit 6 Messern».
  // Steht VOR der Zubehör-Ausnahme, weil «schutzhülle»/«scheide» dort als reines Zubehör gelten.
  const m = HANDKLINGE_MIT_ZUBEHOER.exec(t);
  if (m && !HANDKLINGE_GERAET.test(t)) {
    const vor = t.slice(0, m.index);
    if (KLINGEN_NOMEN.has(m[1].toLowerCase())) return true;
    if (HANDKLINGE_STAMM.test(vor) && !HANDKLINGE_KEIN_PAKET.test(vor)) return true;
  }
  if (HANDKLINGE_KEIN_PAKET.test(t) || HANDKLINGE_GERAET.test(t)) return false;
  return HANDKLINGE_STAMM.test(t);
}
